// @syncfusion/ej2-base/e2e/m.protractor.config.js
var shell = require('shelljs');
var os = require('os');
var camelCase = require('pix-diff/lib/camelCase.js');

exports.config = {

    'seleniumAddress': 'http://hub-cloud.browserstack.com/wd/hub',
    'capabilities': {
        'browserstack.user': process.env.BROWSER_STACK_USER,
        'browserstack.key': process.env.BROWSER_STACK_KEY,
        'device': 'samsung Galaxy S8',
        'browserName': 'chrome',
        'realMobile': 'true',
        'browserstack.debug': 'true',
        'os_version': '7.0'
    },

    allScriptsTimeout: 900000,

    getPageTimeout: 90000,

    framework: 'jasmine',

    jasmineNodeOpts: {
        defaultTimeoutInterval: 1000000
    },

    // Path to change
    specs: ['../../../../e2e/tests/**/*.spec.js'],
    onComplete: function () {
        browser.driver.quit();
    },

    onPrepare: function () {
        const PixDiff = require('pix-diff');
        const fs = require('fs');
        const path = require('path');
        browser.ignoreSynchronization = true;
        browser.waitForAngularEnabled = false;
        browser.isDesktop = false;
        //Path to change
        browser.basePath = require('../../../../protractor.browser.json').basePath;

        // Print User Agent
        browser.executeAsyncScript(function (callback) {
            callback(navigator.userAgent);
        }).then(function (uA) {
            console.log(uA);
        });

        browser.getCapabilities().then(function (cap) {
            browser.browserName = cap.get('browserName');

            browser.pixResult = PixDiff;

            browser.pixDiff = new PixDiff({
                basePath: './e2e',
                diffPath: './e2e',
                formatImageName: '{tag}'
            });
            browser.pixDiff.platformName = cap.get('platformName');
            browser.pixDiff.deviceModel = cap.get('deviceModel');
            browser.pixDiff.deviceName = cap.get('deviceName');
            if (cap.get('platformName') === 'Android') {

                browser.pixDiff.devicePath = browser.pixDiff.platformName + '/' + browser.pixDiff.deviceModel;
            }
            else {
                browser.pixDiff.devicePath = browser.pixDiff.platformName + '/' + browser.pixDiff.deviceName;
            }
            console.log('check Device Path is : ', browser.pixDiff.devicePath);

            //override difference path
            browser.pixDiff.diffPath = path.normalize(camelCase('./e2e/' + browser.pixDiff.devicePath + '/Diff/'));
            //create folder if not present
            createF(browser.pixDiff.diffPath);

            browser.compareScreen = function (element, fileName, opt) {
                createF(camelCase('e2e/expected/' + browser.pixDiff.devicePath));
                createF(camelCase('e2e/actual/' + browser.pixDiff.devicePath));

                // thresholdType: 'percent',
                // threshold: 0.009,
                var option = {
                    imageAPath: '/expected/' + browser.pixDiff.devicePath + '/' + fileName, // Use file-path 
                    imageBPath: '/actual/' + browser.pixDiff.devicePath + '/' + fileName,
                    filter: ['grayScale'],
                    debug: true,
                    hideShift: true,
                };
                var doneFn = arguments[arguments.length - 1];
                if (typeof opt === 'object' && Object.keys(opt).length) {
                    Object.assign(option, opt);
                }
                //Path to change
                if (!fs.existsSync(path.resolve(__dirname, '../../../../' + camelCase('e2e/Expected/' + browser.pixDiff.devicePath + '/' + fileName) + '.png'))) {
                    browser.pixDiff.saveRegion(element, '/Expected/' + browser.pixDiff.devicePath + '/' + fileName, option).then(
                        function () {
                            console.log('Expect Image Created');
                            browser.saveCheckImage(element, fileName, option, doneFn);
                        }
                    );
                } else {
                    browser.saveCheckImage(element, fileName, option, doneFn);
                }

            }
            browser.saveCheckImage = function (element, fileName, option, doneFn) {
                browser.pixDiff.saveRegion(element, '/Actual/' + browser.pixDiff.devicePath + '/' + fileName, option).then(function () {
                    browser.pixDiff.checkRegion(element, '/Expected/' + browser.pixDiff.devicePath + '/' + fileName, option).then(function (result) {
                        //
                        // *  - `RESULT_UNKNOWN`: 0
                        // *  - `RESULT_DIFFERENT`: 1
                        // *  - `RESULT_SIMILAR`: 7
                        // *  - `RESULT_IDENTICAL`: 5
                        expect(result.code).toEqual(browser.pixResult.RESULT_IDENTICAL);
                        if (typeof doneFn === 'function') {
                            doneFn();
                        }
                    });
                });
            }

            browser.waitForEvent = function (id, moduleName, eventName) {
                return browser.executeAsyncScript(function (id, moduleName, eventName, callback) {
                    var instances = document.getElementById(id).ej2_instances;
                    var instanceObj;
                    for (var i = 0; instances && i < instances.length; i++) {
                        if (instances[i].getModuleName() == moduleName) {
                            instanceObj = instances[i]
                        }
                    }
                    if (instanceObj) {
                        var handler;
                        handler = function (e) {
                            instanceObj.removeEventListener(eventName, handler);
                            callback();
                        };
                        instanceObj.addEventListener(eventName, handler);
                    } else {
                        callback();
                    }
                }, id, moduleName, eventName);
            }

            browser.injectScript = function (path) {
                return browser.executeAsyncScript(function (path, callback) {
                    var head = document.getElementsByTagName('head')[0];
                    var script = document.createElement('script');
                    script.type = 'text/javascript';
                    script.onload = function () {
                        callback();
                    }
                    script.src = path;
                    head.appendChild(script);
                }, browser.basePath + path);

            }

            browser.injectCss = function (content) {
                return browser.wait(browser.executeScript(`
                        var style = document.createElement('style');
                        style.id = 'browsercss';
                        if (style.styleSheet) {style.styleSheet.cssText = '` + content + `';}
                        else{style.appendChild(document.createTextNode('` + content + `'));}
                        document.head.appendChild(style);
                        `));
            }

            browser.load = function (path) {
                browser.get(browser.basePath + path);
                browser.wait(function () {
                    browser.sleep(30000);
                    return element(By.css('.e-control')).isDisplayed();
                }, 30000);
                if (browser.css) {
                    browser.injectCss(browser.css);
                }
            }
        });

    },
};

function createF(path) {
    shell.mkdir('-p', path);
}
